Specification for semantic information and CXF

anandkp92 opened the pull request:
* updated specification for including semantic information
* added new page for CXF specification (including classes and properties used in CXF)

anandkp92 commented:
@mwetter can you take a look at the changes to CDL semantic information and the CXF specification?

mwetter commented on specification/source/cdl.rst:
```diff
@@ -2066,18 +2065,23 @@ Example:
+information shall be replaced with the fully qualified path of the instance
+that contains the semantic annotation.  A fully qualified path to an instance
+refers to the complete hierarchical path that specifies the instance's 
+location within an object structure. This qualified path shall include all 
```
This should be *fully* qualified name to use consistent terminology.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -2066,18 +2065,23 @@ Example:
+refers to the complete hierarchical path that specifies the instance's 
+location within an object structure. This qualified path shall include all 
+parent instances leading up to the current instance, with each level of 
+instantiation separated by an underscore (“_”). If an instance is nested
```
How does this affect the translation from CXF to CDL, if the block is in CDL 
`Buildings.Controls.OBC.ASHRAE.G36.VentilationZones.ASHRAE62_1.Setpoints`?
In CXF, with this convention, this becomes `Buildings_Controls_OBC_ASHRAE_G36.VentilationZones_ASHRAE62_1_Setpoints` but there is no package `1` and translating it to CDL would give an error and a different structure.

anandkp92 replied:
So this specification `<instanceName>`, does not affect CXF. 
It only affects the semantic model extracted from a CDL class, and that too only the instance name. The class name, if used in the semantic model, will remain as is, but it's the name of the instance that will be appended with `_`. 

So in your example, if the file structure is 
```
model A
    B controller; 
    ...
end A;
```

```
model B
    Buildings.Controls.OBC.ASHRAE.G36.VentilationZones.ASHRAE62_1.Setpoints heaSet
        annotation(__cdl(semantic(metadataLanguage="Brick 1.3 text/turtle" 
                "bldg:<instanceName> a brick:Heating_Septoint .")));
end B;
```

The semantic model will have the following content:
```
bldg:controller_heaSet a brick:Heating_Setpoint .
```

where `<instanceName>` was replaced by `controller_heaSet`

mwetter commented on specification/source/glossary.rst:
```diff
@@ -52,6 +52,12 @@ the Open Building Controls project.
+   Control eXchange Format
+     The Control eXchange Format (CXF) is a JSON-LD representation
+     of a CDL sequence that can be used to exchange or translate
+     control sequences across platforms. CXF specifications 
```
Don't use "platform" as it means a lot of things. Also, CXF cannot be used to translate (a translator translates, but json-ld is just the input or output format).

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+
+    within ExamplePackage;
+    block ExampleSeq
+    CDL.Reals.MultiplyByParameter gain(k = 100000 
```
Syntax error.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -2066,18 +2065,23 @@ Example:
-This avoids the user having to repeat the name of the instance and
+``<instanceName>``: The text ``<instanceName>`` (including the ``<`` 
+and ``>`` characters) within the metadata of an annotation containing semantic 
+information shall be replaced with the fully qualified path of the instance
```
That is the wrong term. It is called "fully qualified name". See the Modelica Language Definition.

mwetter commented on specification/source/cdl.rst:
```diff
@@ -2105,7 +2109,39 @@ Example:
+shall refer to the name of a parameter instance within the 
+class. The text ``<parameter>`` (including the ``<`` and ``>`` characters) 
+shall be replaced by the value of the parameter. If the parameter does not 
+exist in the class, the ``<parameter>`` shall not be replaced. 
```
The sentence
>  If the parameter does not exist in the class, the ``<parameter>`` shall not be replaced.
sounds to me like someone refers to a parameter name but there is no such parameter. Is this valid? I would think you need to specify "The class must have an instance of a parameter with the name specified by ``<parameter>``, otherwise the specification is not valid.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+  * Blocks that utilize vectors and matrices. These blocks are 
+    used for the development and support of control sequence 
+    libraries but are not easily utilized in graphical 
+    programming tools. 
```
Change
> but are not easily utilized in graphical programming tools

to 

> but may  not be supported by the commercial control product.

(This has nothing to do with graphical programming as Modelica, Simulink etc does graphical programming with matrices.)

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+connectors, parameters, and constants as defined within definition 
+of the Block. To support this process, a Resource Description 
+Framework (RDF) graph representation of the standard has been 
+provided in a CXF-Core.jsonld file using the MIME type 
```
Format at code

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+   ============================  ===========================================================
+   Class                         Description
+   ============================  ===========================================================
+   Package						 
```
what happened to the intentation here?

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+   								 blocks.
+   Blocks   					 A block is the abstract interface of a control sequence.
+   ElementaryBlock 
+   								 An elementary block defined by ASHRAE S231P (subClassOf 
```
Above you use capital letters for Elementary Block, Composite Block etc which is good, but here in this table you use lower case. Change to capital letters to be consistent.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+   Blocks   					 A block is the abstract interface of a control sequence.
+   ElementaryBlock 
+   								 An elementary block defined by ASHRAE S231P (subClassOf 
+   								 Block) (:numref:`sec_ele_blo`).
```
If `subClassOf Block` is code it needs to be formatted as such.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+                                   InputConnector    OutputConnector  to the input of a block. Only connectors
+                                                                      that carry the same data type can be
+                                                                      connected.
+   translationSoftware             Package, Block    String           Used to include the name of the software
+                                                                      used to CXF representation of the
+                                                                      sequence.
+   translationSoftwareVersion      Package, Block    String           Used to include the version of the
+                                                                      software used to CXF representation of
+                                                                      the sequence.		
```
Fix english grammar.

mwetter replied:
@anandkp92 You need to correct the grammar. "Used to include the version of the software used to CXF representation of the sequence" is not English. The same comment applies for the entry above.

Also, make "connectors, Parameters and Constants." consistent. Either use the technical terms (preferred), in which case it is InputConnector and OutputConnector, or use lower case for all.

Moreover, it is not "(local or over network)" but "(local or on the network)" as communication is over the network but the file is on the network.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+                                                                      the sequence.		
+   =============================== ================= ================ =========================================
+
+All the ElementaryBlock within the standard have been 
```
Format ElementaryBlock as code.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+class shall be translated to CXF. If there is no 
+``export`` annotation attached to the instance, the 
+default value would be false and no CXF will be 
+generated for the class of that instance. 
```
Change

>  If there is no 
``export`` annotation attached to the instance, the 
default value would be false and no CXF will be 
generated for the class of that instance. 

to
> Specifying the `export` annotation is optional, and if unspecified `export=false` is assumed.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+        "My gain";
+    end ExampleSeq;
+
+CXF reference to ``gain`` instance: ``ExamplePackage.ExampleSeq#gain``
+CXF reference to ``gain.k`` instance: ``ExamplePackage.ExampleSeq#gain.k``
+CXF property linking ``gain`` and ``gain.k``: ``ExamplePackage.ExampleSeq#gain S231:hasInstance ExamplePackage.ExampleSeq#gain.k .``
```
That won't be rendered properly in html.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+include certain configuration options that specify 
+how the translation will handle:
+
+* Vectors: A tool that generates CXF translations 
```
What about matrices? The same applies but the text is written as if it were for vectors.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+  appended with the underscore (``_``) character. 
+  For matrices (two dimensions) and arrays with more than
+  two dimensions, each index shall be concatenated
+  with the underscore character (``_``).
```
You also need to add in normative text how an assignment is serialized. See https://fmi-standard.org/docs/3.0/ section 2.2.7.1 for the order used to serialize a matrix into a one-dimensional string.

anandkp92 replied:
I'm a bit unclear about this. In CDL, if someone is assigning values to individual elements of a matrix using indices, such as `M[1,2] = a12`, in CXF, wouldn't this become `M_1_2=a12` anyway?

anandkp92 replied:
If the values of an array is being specified in CDL, how will this be specified in CXF? (adhere to FMI)

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+  with an error. If there are parameters that do not 
+  have a value binding, the translation process 
+  shall require the user to input the value for
+  such parameters.
```
You should delete
> If there are parameters that do not 
  have a value binding, the translation process 
  shall require the user to input the value for
  such parameters.

Everything in this sentence has been said in the previous sentence, i.e., it is an error. What the translation process does about it other than issuing an error need not be written here.

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,290 @@
+  expressions in the CDL sequence such as those 
+  within assignment operations, conditional 
+  assignments and arithmetic operations. By default,
+  the expressions shall be preseved in CXF. If the 
```
You write that the expressions are preserved. In other parts you also write that `aa.bb` is converted to `aa_bb`. So if the expression is preserved, it contains a reference to `aa.bb` but if we indeed use underscores (as separate discussion) then we have invalid references.

anandkp92 replied:
As mentioned earlier, the changing the name of instances only happens in the semantic model and not for CXF. The parameters will remain `aa.bb` and the separator will not be replaced with `_`.

mwetter reviewed (changes requested):
@anandkp92 : Please see inline comments.

anandkp92 commented:
@mwetter I've made necessary changes and resolved most of your comments. I do have some questions for the remaining comments.

anandkp92 commented:
@mwetter just a reminder for your review

mwetter commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,288 @@
+include certain configuration options that specify 
+how the translation will handle:
+
+* Vectors: A tool that generates 
+  CXF translations from CDL shall optionally 
+  include a configuration
+  indicating whether or not to flatten or 
+  preserve the vector references. By default, 
+  vector references in CDL should be preserved 
+  in CXF. If vector references should be flattened, 
+  an index appearing within square 
+  brackets (``[`` and ``]``) in CDL shall be 
+  appended with the underscore (``_``) character. 
+  
+  [For example, ``A[1]`` becomes ``A_1``.
+  ]
+* Matrices and arrays with more than
```
I don't think it is helpful to distinguish between vectors and matrices. Why not write these sections in general for 1 and multi-dimensional arrays as they are treated the same.

anandkp92 replied:
what happens if the flattened array instance name is already exisitng as a parameter?

anandkp92 replied:
If the flattened array refernece already exists, Raise an exception

mwetter reviewed (changes requested):
@anandkp92 : Please see inline comments

anandkp92 commented on specification/source/cxf.rst:
```diff
@@ -0,0 +1,288 @@
+  errors. Many commercial control systems define continuous 
+  data types as ``Analog``. 
+  ]
+
```
Add some more discussion on Analog translation.

anandkp92 commented:
@mwetter ready for your review. I believe I've addressed all the comments, replaced CDL sequence with CDL control logic and also removed references to "Analog" data type.
